### ISSUE TITLE
Clean output files before executing task

### ISSUE THREAD

**Issue by lptr** [gradle/gradle#9021]
We should clean the output files and directories of tasks if:

* we are about to either execute the task actions in rebuild mode, but not if we are executing task actions in incremental mode—we already do this when loading from cache,
* there are no overlapping outputs,
* the task doesn't declare the same file as input _and_ output.

This is to ensure tasks don't leave stale output files around.

**Comment by wolfs** [comment]
I sent the document around. Let's move this back to the backlog until we actually want to do something about it.

**Comment by lptr** [comment]
This is something we still want to do.

**Comment by JojOatXGME** [comment]
Is there any option to declare a task as non-incremental, so that the output directory is always cleaned up prior to the execution? It would be nice if people wouldn't have to implement their own cleanup at the beginning of each task, as this is something which can easily be forgotten without resulting in an immediate error.

**Comment by lptr** [comment]
Agreed. There is no way currently to declare the task as non-incremental, unfortunately.

**Comment by nbrugger-tgm** [comment]
Can you until it is implemented, post the "best" way to do this? Would it be a 
```kotlin
doFirst {
    delete(this.output.files)
}
```
?
